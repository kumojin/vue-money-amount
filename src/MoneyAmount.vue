<template>
  <span :class="classes" v-text="formattedAmount" />
</template>

<script>
import { getUserLocale } from 'get-user-locale';
import Helper from '@/helpers';

export default {
  name: 'MoneyAmount',
  props: {
    amount: {
      type: Number,
      required: true,
    },
    currency: {
      type: String,
      required: true,
    },
    isFractionated: {
      type: Boolean,
      required: false,
      default: true,
    },
    locale: {
      type: String,
      required: false,
    },
    minimumFractionDigits: {
      type: Number,
      required: false,
      default: 0,
    },
  },
  computed: {
    classes () {
      if (this.amount > 0) {
        return 'money-amount--positive'
      } else if (this.amount < 0)Â {
        return 'money-amount--negative'
      } else {
        return 'money-amount--zero'
        
      }
    },
    formattedAmount() {
      const options = {
        isFractionated: this.isFractionated,
        locale: this.locale || getUserLocale(),
        minimumFractionDigits: this.minimumFractionDigits,
      };

      return Helper.formatMoneyAmount(this.amount, this.currency, options);
    },
  },
};
</script>
<style lang="scss">
  .money-amount {
    &--positive {
      color: green;
    }

    &--negative {
      color: red;
    }
  }
</style>
