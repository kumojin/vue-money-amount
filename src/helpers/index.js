import { getUserLocale } from 'get-user-locale'
import { utils as currencyUtils } from 'iso4217'

/**
 * Formats the given money "amount" with "currency" according to the
 * locale of the user.
 *
 * Amounts can come as fractionated ("isFractionated" to TRUE), which
 * means they do not include the fraction (ex: USD$4200 is in fact
 * USD$42.00 because the USD currency allows for 2 decimal places)
 *
 * Then we use the standard Intl.NumberFormat to format the unit
 * to the currency using the user's locale.
 *
 * @param {*} amount the money amount
 * @param {*} currency the currency of the amount
 * @param {*} locale the locale of the user, defaults to #getUserLocale
 * @param {*} isFractionated whether the amount comes without decimals, and
 * the decimal point must be determined according to the currency. Defaults to false
 * @param {*} minimumFractionDigits how many fraction digits to show (see `Intl.NumberFormat` options)
 */
export const formatMoneyAmount = (
  amount,
  currency,
  {
    locale = getUserLocale(),
    isFractionated = false,
    minimumFractionDigits = 0
  } = {}
) => {
  if (!currency) {
    return ''
  }

  const precision = isFractionated ? currencyUtils.getByCode(currency)?.Fraction || 0 : 0

  return Intl.NumberFormat(locale, {
    currency,
    style: 'currency',
    minimumFractionDigits // So we don't show .00 if the amount has no decimals
  }).format(amount / Math.pow(10, precision))
}
