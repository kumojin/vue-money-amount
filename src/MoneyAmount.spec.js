import { createLocalVue, shallowMount } from '@vue/test-utils'
import { getUserLocale } from 'get-user-locale'
import { formatMoneyAmount } from './helpers'

import MoneyAmount from './MoneyAmount'

jest.mock('get-user-locale')
jest.mock('./helpers')

const localVue = createLocalVue()

describe('<MoneyAmount />', () => {
  const AMOUNT = 4284
  const CURRENCY = 'USD'
  const CA_LOCALE = 'fr-CA'
  const propsData = {
    amount: AMOUNT,
    currency: CURRENCY
  }
  let wrapper

  const mount = (props) => {
    return shallowMount(MoneyAmount, {
      localVue,
      propsData: {
        ...propsData,
        ...props
      }
    })
  }

  beforeEach(() => {
    getUserLocale.mockReturnValue(CA_LOCALE)
    formatMoneyAmount.mockImplementation((amount, currency, opts) => {
      const optsAsString = Object.keys(opts).reduce((acc, optionName) => {
        acc += `${optionName}: ${opts[optionName]}, `
        return acc
      }, '')
      return `${amount}-${currency}-${optsAsString}`
    })
  })

  it('renders with default properties', () => {
    wrapper = mount()

    expect(wrapper.text()).toBe('4284-USD-isFractioned: true, locale: fr-CA, minimumFractionDigits: 0,')
  })

  it('renders with custom properties', () => {
    wrapper = mount({
      locale: 'fr-FR',
      isFractioned: false,
      minimumFractionDigits: 4
    })

    expect(wrapper.text()).toBe('4284-USD-isFractioned: false, locale: fr-FR, minimumFractionDigits: 4,')
  })
})
