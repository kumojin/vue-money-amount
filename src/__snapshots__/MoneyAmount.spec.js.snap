// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MoneyAmount /> when mounted renders the money amount with default options 1`] = `<span class="money-amount">4284-USD-isFractionated: true, locale: fr-CA, minimumFractionDigits: 0, </span>`;
